samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,253 @@
+
+LOG = logging.getLogger(__name__)
+
+REMOVAL_MAP = {
+    AWS_SERVERLESS_FUNCTION: ["CodeUri", "ImageUri"],
+    AWS_LAMBDA_FUNCTION: {"Code": ["ImageUri", "S3Bucket", "S3Key", "S3ObjectVersion"]},
+    AWS_SERVERLESS_LAYERVERSION: ["ContentUri"],
+    AWS_LAMBDA_LAYERVERSION: ["Content"],
+    AWS_SERVERLESS_API: ["DefinitionBody"],
+    AWS_APIGATEWAY_RESTAPI: ["BodyS3Location"],
+    AWS_SERVERLESS_HTTPAPI: ["DefinitionUri"],
+    AWS_APIGATEWAY_V2_API: ["BodyS3Location"],
+    AWS_SERVERLESS_STATEMACHINE: ["DefinitionUri"],
+    AWS_STEPFUNCTIONS_STATEMACHINE: ["DefinitionS3Location"],
+}
+
+
+class InfraSyncExecutor:
+    """
+    Executor for infra sync that contains skip logic when template is not changed
+    """
+
+    _build_context: BuildContext
+    _package_context: PackageContext
+    _deploy_context: DeployContext
+
+    def __init__(self, build_context: BuildContext, package_context: PackageContext, deploy_context: DeployContext):
+        """Constructs the sync for infra executor.
+
+        Parameters
+        ----------
+        build_context : BuildContext
+            BuildContext
+        package_context : PackageContext
+            PackageContext
+        deploy_context : DeployContext
+            DeployContext
+        """
+        self._build_context = build_context
+        self._package_context = package_context
+        self._deploy_context = deploy_context
+
+        session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
+        self._cfn_client = session.client("cloudformation")
+        self._s3_client = session.client("s3")
+
+    def _compare_templates(self, local_template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        local_template_path : str
+            The local template location
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        try:
+            current_template = None
+            # If the customer template uses a nested stack with location/template URL in S3
+            if local_template_path.startswith("https://"):
+                parsed_s3_location = re.search(r"https:\/\/[^/]*\/([^/]*)\/(.*)", local_template_path)
+                if parsed_s3_location:
+                    s3_bucket = parsed_s3_location.group(1)
+                    s3_key = parsed_s3_location.group(2)
+                    s3_object = self._s3_client.get_object(Bucket=s3_bucket, Key=s3_key)
+
+                    streaming_body = s3_object.get("Body")
+                    if streaming_body:
+                        current_template = yaml_parse(streaming_body.read().decode("utf-8"))
```

Comment:
Do we still need to pull a template from remote given that we moved comparison logic before package and after build?

Reply:
Yes we do, again nested stack location can be remote before we even build

Reply:
I've started reviewing this early in the morning and somehow this comment made it to afternoon session even though it wasn't there when I was publishing my review. Apologies for duplication.